File: ocr/provider.go
package ocr

import (
	"context"
	"io"

	"doctor-service/entities"
)

type ExtractedFields struct {
	DoctorName         string  `json:"doctor_name"`
	RegistrationNumber string  `json:"registration_number"`
	RegistrationCouncil string `json:"registration_council"`
	RegistrationYear   int     `json:"registration_year"`
	Degree             string  `json:"degree"`
	Specialization     string  `json:"specialization"`
	University         string  `json:"university"`
	College            string  `json:"college"`
	YearCompleted      int     `json:"year_completed"`
	DOB                string  `json:"dob"`
	GovtIDNumber       string  `json:"govt_id_number"`
}

type OCRResult struct {
	RawJSON          string
	ParsedFields     ExtractedFields
	Confidence       float64
	ProcessingTimeMS int64
}

type OCRProvider interface {
	Extract(ctx context.Context, doc *entities.DoctorDocument, fileReader io.Reader) (*OCRResult, error)
	GetProviderName() string
}

File: python_backend/ocr/provider.py
from abc import ABC, abstractmethod

class ExtractedFields:
    def __init__(self, doctor_name="", registration_number="", registration_council="", registration_year=0, degree="", specialization="", university="", college="", year_completed=0, dob="", govt_id_number=""):
        self.doctor_name = doctor_name
        self.registration_number = registration_number
        self.registration_council = registration_council
        self.registration_year = registration_year
        self.degree = degree
        self.specialization = specialization
        self.university = university
        self.college = college
        self.year_completed = year_completed
        self.dob = dob
        self.govt_id_number = govt_id_number

class OCRResult:
    def __init__(self, raw_json, parsed_fields, confidence, processing_time_ms):
        self.raw_json = raw_json
        self.parsed_fields = parsed_fields
        self.confidence = confidence
        self.processing_time_ms = processing_time_ms

class OCRProvider(ABC):
    @abstractmethod
    def extract(self, context, doc, file_reader) -> OCRResult:
        pass

    @abstractmethod
    def get_provider_name(self) -> str:
        pass
